scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala

```diff
@@ -0,0 +1,153 @@
+package com.twitter.scalding.parquet.scrooge
+
+import java.util
+
+import org.apache.parquet.schema.Type.Repetition
+import org.apache.parquet.schema.{GroupType, MessageType, Type}
+import org.apache.parquet.thrift.DecodingSchemaMismatchException
+import org.slf4j.LoggerFactory
+
+import scala.collection.mutable
+
+object ParquetCollectionFormatForwardCompatibility {
+
+  private val LOGGER = LoggerFactory.getLogger(getClass)
+
+  /**
+   * Create a forward-compatible schema, using content from source type with format from target type.
+   * @param sourceType source type with legacy format
+   * @param targetType target type to which source is converted to
+   */
+  def formatForwardCompatibleMessage(sourceType: MessageType, targetType: MessageType): MessageType = {
+    val groupResult = formatForwardCompatibleType(sourceType, targetType).asGroupType()
+    LOGGER.info("Making source schema to be compatible with target" +
+      s"\nSource:\n${sourceType}\nTarget:\n${targetType}\nResult:\n${groupResult}")
+    new MessageType(groupResult.getName, groupResult.getFields)
+  }
+
+  /**
+   * Traverse source/target schemas and format nodes of list or map.
+   * The formatting is not to one-to-one node swapping from source to target,
+   * this is because the subset fields of source node and its optional/required must
+   * be maintained in the formatted result.
+   */
+  private def formatForwardCompatibleType(sourceType: Type, targetType: Type): Type = {
```

Comment:
I find this method's pattern match to be a bit difficult to intuitively understand and I think that maybe it's because of the `GroupUnwrapped`: there, we seem to have a big cross product of potential possibilities, but it looks like we only have explicit cases for some of them.

By construction, it looks like we never wind up with `repeatedListType` and `repeatedMapType` both non-None.

To simplify the pattern match and `unwrapGroup`, what do you think about instead defining a `sealed trait` of case-classes corresponding to the different cases that we might need to handle? What that approach, I think we can avoid the need to do this more complicated positional pattern-matching. 

Given that, could it make sense to have methods like `ParquetMapFormatRule.isGroupMap()` instead be "extractor"-type methods which return `Option[GroupMapSomethingIHaventNamedYet]`? That could help to avoid separate definition and extraction code that we have now (where the "is X" and "extract X" code for the same logical concept are in separate places).
 
This would also have the advantage of avoiding the need to have comments like `// Format list` to explain the cases, since that semantic information could be instead carried in the sealed trait class names.